src/eos-updater-util.c

```diff
@@ -313,73 +323,193 @@ eos_updater_resolve_upgrade (EosUpdater  *updater,
       g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
                    "Failed to download branch config data (HTTP %d),"
                    " cannot upgrade", status);
-      goto out;
+      return NULL;
     }
 
   bkf = g_key_file_new ();
   if (!g_key_file_load_from_data (bkf, msg->response_body->data, -1,
                                   G_KEY_FILE_NONE, error))
-    goto out;
+    return NULL;
+  return g_steal_pointer (&bkf);
+}
+
+static gboolean
+process_single_group (GKeyFile *bkf,
+                      const gchar *group_name,
+                      gboolean *on_hold,
+                      gchar **p_ref,
+                      GError **error)
+{
+  GError *local_error = NULL;
+
+  if (g_key_file_get_boolean (bkf, group_name, ON_HOLD_KEY, &local_error))
+    {
+      *on_hold = TRUE;
+      return TRUE;
+    }
+  /* The "OnHold" key is optional. */
+  if (!g_error_matches (local_error,
+                        G_KEY_FILE_ERROR,
+                        G_KEY_FILE_ERROR_KEY_NOT_FOUND))
+    {
+      g_propagate_error (error, local_error);
+      return FALSE;
+    }
+  g_clear_error (&local_error);
+  *p_ref = g_key_file_get_string (bkf, group_name, OSTREE_REF_KEY, error);
+  if (*p_ref == NULL)
+    return FALSE;
+  return TRUE;
```

Comment:
I think you should set `*on_hold = FALSE` before returning `TRUE` here. When the error is not set, all the out parameters should always be filled.